src/rust/engine/options/src/parse.rs

```diff
@@ -274,49 +278,188 @@ fn format_parse_error(
+                        Err(err) if err.kind() == io::ErrorKind::NotFound => Ok((Some(path), None)),
+                        Err(err) => Err(mk_parse_err(err, &path)),
+                    }
+                }
+                _ => {
+                    let path = PathBuf::from(suffix);
+                    let content = fs::read_to_string(&path).map_err(|e| mk_parse_err(e, &path))?;
+                    Ok((Some(path), Some(content)))
+                }
+            }
+        }
+    } else {
+        Ok((None, Some(value)))
+    }
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_int_list(value: &str) -> Result<Vec<ListEdit<i64>>, ParseError> {
-    option_value_parser::int_list_edits(value).map_err(|e| format_parse_error("int list", value, e))
+pub(crate) fn expand(value: String) -> Result<Option<String>, ParseError> {
+    let (_, expanded_value) = maybe_expand(value)?;
+    Ok(expanded_value)
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_float_list(value: &str) -> Result<Vec<ListEdit<f64>>, ParseError> {
-    option_value_parser::float_list_edits(value)
-        .map_err(|e| format_parse_error("float list", value, e))
+pub(crate) fn expand_to_list<T: ListMember>(
+    value: String,
+) -> Result<Option<Vec<ListEdit<T>>>, ParseError> {
+    let (path_opt, value_opt) = maybe_expand(value)?;
+    if value_opt.is_none() {
+        return Ok(None);
+    }
+    let value = value_opt.unwrap();
+    let mut deserialized_items: Option<Vec<T>> = None;
+    if let Some(path) = path_opt {
+        if let Some(ext) = path.extension() {
+            if ext == "json" {
+                deserialized_items =
+                    Some(serde_json::from_str(&value).map_err(|e| mk_parse_err(e, &path))?)
+            } else if ext == "yml" || ext == "yaml" {
+                deserialized_items =
+                    Some(serde_yaml::from_str(&value).map_err(|e| mk_parse_err(e, &path))?)
+            }
+        }
+    }
+    if let Some(items) = deserialized_items {
+        Ok(Some(vec![ListEdit {
+            action: ListEditAction::Replace,
+            items,
+        }]))
+    } else {
+        T::parse_list(&value).map(Some)
+    }
 }
 
-pub(crate) fn parse_string_list(value: &str) -> Result<Vec<ListEdit<String>>, ParseError> {
-    option_value_parser::string_list_edits(value)
-        .map_err(|e| format_parse_error("string list", value, e))
+pub(crate) fn expand_to_dict(value: String) -> Result<Option<DictEdit>, ParseError> {
+    let (path_opt, value_opt) = maybe_expand(value)?;
+    if value_opt.is_none() {
+        return Ok(None);
+    }
+    let value = value_opt.unwrap();
+    let mut deserialized_items: Option<HashMap<String, Val>> = None;
+    if let Some(path) = path_opt {
+        if let Some(ext) = path.extension() {
+            if ext == "json" {
+                deserialized_items =
+                    Some(serde_json::from_str(&value).map_err(|e| mk_parse_err(e, &path))?)
```

Comment:
There's a little bit of repetition here vis the previous function, but the differences are just subtle enough that it became more confusing than it was worth to try and unify them.

Reply:
What do you think about at least pulling out the file-type-identification? Maybe something like

```rust
enum FileType {
    Json,
    Yalm
}

impl FileType {
    fn guess(path: &Path) -> Option<FileType> {
        match path.extension() {
            "json" => Some(FileType::Json),
            "yaml" | "yml" => ..., 
            _ => None,
        }
    }
}
```

and then this section of the code in both could could be something like:

```rust
let deserialized_items = match path_opt.flat_map(|p| FileType::guess(&p)) {
   Some(FileType::Json) => ...,
   Some(FileType::Yaml) => ...,
   None => None,
};
```

One could potentially then even move the "call `serde_json` or `serde_yaml`" decision to a generic method on `FileType`, and have that be shared.

Reply:
I toyed with this some more and was able to create a generic deserialization helper function that works for lists and dicts, so hooray Rust type system!